controllers/front/callbacknotification.php

```diff
@@ -22,124 +22,151 @@ public function postProcess()
-                        $sql = 'UPDATE `' . _DB_PREFIX_ . 'altapay_order` 
-                        SET `paymentStatus` = \'declined\' WHERE `id_order` = ' . $order->id;
-                        Db::getInstance()->Execute($sql);
-                        $this->unlock($fp);
-                        exit('Order status updated to Error');
-                    } else {
-                        // Unexpected scenario
-                        $mNa = $this->module->name;
-                        PrestaShopLogger::addLog('Unexpected scenario: Callback notification was received for Transaction '
-                                                 . $shopOrderId . ' with payment status ' . $transactionStatus, 3, '1005', $mNa,
-                            $this->module->id, true);
-                        $this->unlock($fp);
-                        exit('Unrecognized status received ' . $transactionStatus);
-                    }
+    public function handleNotificationAction($cart, $order, $response, $customer, $transactionStatus, $shopOrderId)
+    {
+        // NO ORDER FOUND, CREATE?
+        if (!Validate::isLoadedObject($order)) {
+            // Payment successful - create order
+            if ($response && is_array($response->Transactions)) {
+                $order_status = (int) Configuration::get('PS_OS_PAYMENT');
+                $currency_paid = Currency::getIdByIsoCode($response->Currency);
+                $amount_paid = $response->amount;
+                $paymentType = $response->Transactions[0]->AuthType;
+                /* If payment type is 'payment' funds have not yet been captured,
+                * so AltaPay returns 0 as the captured amount.Therefore we assume full payment has been authorized.
+                */
+                if ($paymentType === 'payment') {
+                    $amount_paid = $cart->getOrderTotal(true, Cart::BOTH);
+                    $currency_paid = new Currency($cart->id_currency);
+                }
+                // Determine payment method for display
+                $paymentMethod = determinePaymentMethodForDisplay($response);
+
+                // Create an order with 'payment accepted' status
+                $cSk = $customer->secure_key;
+                $cpId = (int) $currency_paid->id;
+                $cId = $cart->id;
+                $oSt = $order_status;
+                $pMeth = $paymentMethod;
+                $this->module->validateOrder($cId, $oSt, $amount_paid, $pMeth, null, null, $cpId, false, $cSk);
+                // Log order
+                $current_order = new Order((int) $this->module->currentOrder);
+
+                createAltapayOrder($response, $current_order, $transactionStatus);
+                exit('Order created');
+            } else {
+                exit('Only handling Success state');
+            }
+        } else {
+            if ($transactionStatus === 'preauth' || $transactionStatus === 'bank_payment_finalized' || $transactionStatus === 'captured') {
```

Comment:
If we know that the order has been call with result Successful, is it advised to check on the status?

You are missing: invoice_initialized, recurring_confirmed

Reply:
Removed unnecessary check